disttask/framework/dispatcher/dispatcher.go

```diff
@@ -74,211 +63,233 @@ type dispatcher struct {
+	logPrefix := fmt.Sprintf("task_id: %d, task_type: %s", task.ID, task.Type)
 	return &dispatcher{
 		ctx,
 		taskMgr,
 		task,
+		logutil.WithKeyValue(context.Background(), "dispatcher", logPrefix),
 	}
 }
 
-// ExecuteTask start to schedule a task
-func (d *dispatcher) ExecuteTask() {
-	logutil.BgLogger().Info("execute one task", zap.Int64("task ID", d.task.ID),
+// ExecuteTask start to schedule a task.
+func (d *dispatcher) executeTask() {
+	logutil.Logger(d.logCtx).Info("execute one task",
 		zap.String("state", d.task.State), zap.Uint64("concurrency", d.task.Concurrency))
-	d.scheduleTask(d.task.ID)
+	d.scheduleTask()
 }
 
-// monitorTask checks whether the current step of one task is finished,
-// and gather subTaskErrs to handle subTask fails.
-func (d *dispatcher) monitorTask(taskID int64) (finished bool, subTaskErrs []error) {
-	// TODO: Consider putting the following operations into a transaction.
-	var err error
-	d.task, err = d.taskMgr.GetGlobalTaskByID(taskID)
+// monitorTask fetch task state from tidb_global_task table.
+func (d *dispatcher) monitorTask() (err error) {
+	d.task, err = d.taskMgr.GetGlobalTaskByID(d.task.ID)
 	if err != nil {
-		logutil.BgLogger().Error("check task failed", zap.Int64("task ID", d.task.ID), zap.Error(err))
-		return false, nil
-	}
-	switch d.task.State {
-	case proto.TaskStateCancelling:
-		return false, []error{errors.New("cancel")}
-	case proto.TaskStateReverting:
-		cnt, err := d.taskMgr.GetSubtaskInStatesCnt(d.task.ID, proto.TaskStateRevertPending, proto.TaskStateReverting)
-		if err != nil {
-			logutil.BgLogger().Warn("check task failed", zap.Int64("task ID", d.task.ID), zap.Error(err))
-			return false, nil
-		}
-		return cnt == 0, nil
-	default:
-		subTaskErrs, err = d.taskMgr.CollectSubTaskError(d.task.ID)
-		if err != nil {
-			logutil.BgLogger().Warn("collect subtask error failed", zap.Int64("task ID", d.task.ID), zap.Error(err))
-			return false, nil
-		}
-		if len(subTaskErrs) > 0 {
-			return false, subTaskErrs
-		}
-		// check subtasks pending or running.
-		cnt, err := d.taskMgr.GetSubtaskInStatesCnt(d.task.ID, proto.TaskStatePending, proto.TaskStateRunning)
-		if err != nil {
-			logutil.BgLogger().Warn("check task failed", zap.Int64("task ID", d.task.ID), zap.Error(err))
-			return false, nil
-		}
-		return cnt == 0, nil
+		logutil.Logger(d.logCtx).Error("monitor task failed", zap.Error(err))
 	}
+	return err
 }
 
 // scheduleTask schedule the task execution step by step.
-func (d *dispatcher) scheduleTask(taskID int64) {
+func (d *dispatcher) scheduleTask() {
 	ticker := time.NewTicker(checkTaskFinishedInterval)
 	defer ticker.Stop()
 	for {
 		select {
 		case <-d.ctx.Done():
-			logutil.BgLogger().Info("schedule task exits", zap.Int64("task ID", taskID), zap.Error(d.ctx.Err()))
+			logutil.Logger(d.logCtx).Info("schedule task exits", zap.Error(d.ctx.Err()))
 			return
 		case <-ticker.C:
-			stepIsFinished, errs := d.monitorTask(taskID)
+			err := d.monitorTask()
```

Comment:
The name of this function is misleading. At first glance I think there is a new go-routine in it.